tests/std/tests/P2474R2_views_repeat/test.cpp

```diff
@@ -0,0 +1,294 @@
+        static_assert(is_nothrow_move_constructible_v<I>);
+
+        tmp = i;
+        static_assert(is_nothrow_copy_assignable_v<I>);
+
+        i = move(tmp);
+        static_assert(is_nothrow_move_assignable_v<I>);
+    }
+
+    assert(!(first == second));
+    assert(!(second == first));
+    static_assert(noexcept(first == second)); // strengthened
+
+    assert(first != second);
+    assert(second != first);
+    static_assert(noexcept(first != second)); // strengthened
+
+    assert(first < second);
+    assert(!(second < first));
+    static_assert(noexcept(first < second)); // strengthened
+
+    assert(!(first > second));
+    assert(second > first);
+    static_assert(noexcept(first > second)); // strengthened
+
+    assert(first <= second);
+    assert(!(second <= first));
+    static_assert(noexcept(first <= second)); // strengthened
+
+    assert(!(first >= second));
+    assert(second >= first);
+    static_assert(noexcept(first >= second)); // strengthened
+
+    assert(first <=> second < 0);
+    assert(second <=> first > 0);
+    static_assert(noexcept(first <=> second)); // strengthened
+
+    {
+        I i = first;
+
+        assert(&++i == &i);
+        assert(i == second);
+        static_assert(noexcept(++i)); // strengthened
+
+        i = first;
+
+        assert(i++ == first);
+        assert(i == second);
+        static_assert(noexcept(i++)); // strengthened
+
+        assert(&--i == &i);
+        assert(i == first);
+        static_assert(noexcept(--i)); // strengthened
+
+        i = second;
+
+        assert(i-- == second);
+        assert(i == first);
+        static_assert(noexcept(i--)); // strengthened
+
+        assert(i + 1 == second);
+        static_assert(noexcept(i + 1)); // strengthened
+
+        assert(1 + i == second);
+        static_assert(noexcept(1 + i)); // strengthened
+
+        assert(&(i += 1) == &i);
+        assert(i == second);
+        static_assert(noexcept(i += 1)); // strengthened
+        assert(i - 1 == first);
+        static_assert(noexcept(i - 1)); // strengthened
+
+        assert(&(i -= 1) == &i);
+        assert(i == first);
+        static_assert(noexcept(i -= 1)); // strengthened
+
+        assert(second - first == 1);
+        static_assert(noexcept(second - first)); // strengthened
+
+        assert(first[1] == *second);
+        assert(second[-1] == *first);
+        static_assert(noexcept(first[1])); // strengthened
+    }
+
+    const same_as<ranges::sentinel_t<R>> auto last = rng.end();
+    assert(!(first == last));
+    static_assert(noexcept(first == last)); // strengthened
+    assert(first != last);
+    static_assert(noexcept(first != last)); // strengthened
+    if constexpr (bounded) {
+        assert(cmp_equal(last - first, rng.size()));
+        static_assert(noexcept(last - first)); // strengthened
+    }
+}
+
+struct move_tester {
+    int x         = -1;
+    move_tester() = default;
+    constexpr move_tester(move_tester&& other) noexcept {

```

Comment:
Would anyone mind telling me why `noexcept` was necessary here for the `CanTakeDrop` test below?

https://github.com/microsoft/STL/blob/3b5c7e586075c3e320a4280baba73b705c78a001/tests/std/tests/P2474R2_views_repeat/test.cpp#L275-L276

```
Standard Output:
--
test.cpp
K:\MicrosoftSTL\out\build\x64\out\inc\ranges(2967): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'std::ranges::repeat_view<_UnrefU,__int64>'
K:\MicrosoftSTL\tests\std\tests\P2474R2_views_repeat\test.cpp(276): error C2607: static assertion failed
--
```

The test seems to pass without it in a separate project 😕 .

Reply:
I investigated this and I think I'm 80% of the way to the root cause, but I don't fully understand it yet.

Here is where the behavioral difference appears to happen:

```
D:\GitHub\STL\out\x64>type meow.cpp
```
```cpp
#include <ranges>
using namespace std;
using namespace std::ranges;

struct Nonthrowing {
    Nonthrowing() = default;
    constexpr Nonthrowing(Nonthrowing&&) noexcept {}
};

struct Throwing {
    Throwing() = default;
    constexpr Throwing(Throwing&&) {}
};

static_assert(!viewable_range<const repeat_view<Nonthrowing>&>);
static_assert(viewable_range<const repeat_view<Throwing>&>);

int main() {}
```
```
D:\GitHub\STL\out\x64>cl /EHsc /nologo /W4 /std:c++latest meow.cpp
meow.cpp

D:\GitHub\STL\out\x64>clang-cl /EHsc /nologo /W4 /std:c++latest meow.cpp

D:\GitHub\STL\out\x64>
```

`viewable_range` is important because `_Take_fn::operator()` is constrained for `viewable_range _Rng`:
https://github.com/microsoft/STL/blob/2f8342a3a57fb157d881c6a2d42a917d20e413f8/stl/inc/ranges#L2758-L2759

Therefore, when `move_tester`/`Nonthrowing` have a `noexcept` move constructor, which (through a path I haven't fully understood yet) causes `!viewable_range<const repeat_view<Nonthrowing>&>` to be the case, then the `CanTakeDrop` concept can immediately answer `false`.

When `Throwing` has a throwing move constructor, through a path I don't fully understand yet either, `viewable_range<const repeat_view<Throwing>&>` is `true`. Therefore, `CanTakeDrop` proceeds to consider what `_Take_fn::operator()` would do, and that eventually tries to copy a `Throwing` (because `forward_like` ends up forming a `const Throwing&`), and that fails to compile in a non-SFINAE context. That part I understand.

So the remaining question is why `viewable_range` is sensitive to the `noexcept`.

Reply:
A copy constructor of the primary template of `_Movable_box` is [unconstrained](https://github.com/microsoft/STL/blob/2f8342a3a57fb157d881c6a2d42a917d20e413f8/stl/inc/ranges#L275-L279). Therefore `_Movable_box<Throwing>`, and hence `repeat_view<Throwing>`, appears to be copyable.

Reply:
> A copy constructor of the primary template of `_Movable_box` is [unconstrained](https://github.com/microsoft/STL/blob/2f8342a3a57fb157d881c6a2d42a917d20e413f8/stl/inc/ranges#L275-L279). 

Great failure.

Reply:
#3171 should fix the issue.

Reply:
Thanks @SuperWig, @timsong-cpp, and @CaseyCarter for noticing, finding, and fixing the problem here! :heart_eyes_cat: